### DISCUSSION TITLE
TS cache.readFragment could have a wrong type return

### DISCUSSION THREAD

**Discussion by albertcito** [urql-graphql/urql#1488] [Q&A] [answered]
The `rePost` variable has all the `RePostFieldsFragment` values:

```javascript
const rePost = cache.readFragment(
  RePostFieldsFragmentDoc,
  { id: rePostID },
)
```
But the type says it only has `id` value o `null`
<img width="392" alt="Screen Shot 2021-03-29 at 7 51 30 AM" src="https://user-images.githubusercontent.com/1768483/112832631-94f24700-9063-11eb-8ce9-9eba336796a0.png">

Looks like it's because this type:

```javascript 
readFragment<T = Data, V = Variables>(
  fragment: DocumentNode | TypedDocumentNode<T, V>, 
  entity: string | Data | T, 
  variables?: V
): T | null;
```

**Comment by kitten** [comment, marked as answer]
That seems normal depending on what you've passed in. Assuming that that document isn't typed then the default return value is `Data` which only guarantees the system fields of types. It's a little suspicious that it's incomplete but either you'll get `Data`, or if you're generating a typed fragment (which afaik no tool supports properly yet) you'll actually see your full type piped through

**Reply by albertcito** [reply]
Thank you @kitten. 

I added the `as` with the return expected  to the code and it works perfect:

```typescript
const rePost = cache.readFragment(
  RePostFieldsFragmentDoc,
  { id: rePostID },
) as RePostFieldsFragment | null;
```

**Reply by lukasluecke** [reply]
Using a typed fragment actually works as expected, without any further type annotations (generated using the `typed-document-node` plugin in grapqhl-code-generator).

<img width="746" alt="CleanShot 2021-04-01 at 10 46 34@2x" src="https://user-images.githubusercontent.com/6652670/113276254-8f9d3280-92d7-11eb-90e1-f1d09e318e29.png">

<img width="946" alt="CleanShot 2021-04-01 at 10 46 58@2x" src="https://user-images.githubusercontent.com/6652670/113276292-9c218b00-92d7-11eb-9452-aa1e4e345fb0.png">
